First version to CR

rlogic opened the pull request:
- Three types of alert
- Copying predefined config from example
- Using helper

mpptx committed:
First version:
- Three types of alert
- Copying predefined config from example
- Using helper

mpptx committed:
First version -- pre-CR changes:
* Removing Gemfile.lock,
* Adding dependencies to gemspec,
* Replacing .slim with .erb.

rlogic commented on .ruby-version:
```diff
@@ -0,0 +1 @@
+ruby-2.4.6
```
I haven't seen this file in any other gem. Do we need this?

rlogic commented on Gemfile:
```diff
@@ -12,3 +12,5 @@ gemspec
 # To use a debugger
 # gem 'byebug', group: [:development, :test]
+
+gem 'rubocop', '0.66'
```
remove from here and add  to difficult-customer.gemspec: `spec.add_development_dependency "rubocop", "0.66"`

rlogic commented on Gemfile.lock:
```diff
@@ -0,0 +1,137 @@
+PATH
```
As far as I know there should not be any Gemfile.lock in a gem project

rlogic commented on app/helpers/difficult_customer_helper.rb:
```diff
@@ -0,0 +1,28 @@
+    # render_partial 'alert'
+  end
+
+  def difficult_customer_options
```
This should be a private method, right?

rlogic commented on app/helpers/difficult_customer_helper.rb:
```diff
@@ -0,0 +1,28 @@
+
+  private
+
+  def render_partial(partial)
```
kind of a overkill

rlogic commented on app/helpers/difficult_customer_helper.rb:
```diff
@@ -0,0 +1,28 @@
+  end
+
+  def fake_class(string_length: 8)
+    ''.tap { |s| string_length.times { s << ('a'..'z').to_a.sample } }
```
`string_length.times.map { ('a'..'z').to_a.sample }.join`

rlogic commented on app/helpers/difficult_customer_helper.rb:
```diff
@@ -0,0 +1,28 @@
+# Rails helper
+module DifficultCustomerHelper
+  def difficult_customer
+    render_partial %w[alert modal page].sample
```
add a TODO or INFO explaining why it's done this way

rlogic commented on app/helpers/difficult_customer_helper.rb:
```diff
@@ -0,0 +1,28 @@
+    DifficultCustomer::Config.new.config
+  end
+
+  def fake_classes_array(array_length, string_length: 8)
```
`random_dom_ids` :), seriously

rlogic replied:
rename array_length to count and string_length to complexity, also change the way how you call the method:

```
def random_dom_ids(count:, complexity: 10)
  Array.new(count) { random_string(complexity: complexity) }
end

private

def random_string(complexity: complexity)
  complexity.times.map { ('a'..'z').to_a.sample }.join
end
```

rlogic commented on difficult_customer.gemspec:
```diff
@@ -0,0 +1,28 @@
+  s.add_dependency 'bundler', '~> 1.3'
+  s.add_dependency 'rails'
+  s.add_dependency 'rake'
+end
```
as I said use also `add_development_dependency `

rlogic commented on lib/difficult_customer/config.rb:
```diff
@@ -0,0 +1,28 @@
+        'difficult_customer.yml'
+      )
+    end
+
```
remove blank line

rlogic commented on lib/difficult_customer/engine.rb:
```diff
@@ -0,0 +1,9 @@
+module DifficultCustomer
+  # A Rails Engine used make the gem running.
+  class Engine < ::Rails::Engine
+
```
remove

rlogic commented on lib/difficult_customer/engine.rb:
```diff
@@ -0,0 +1,9 @@
+  class Engine < ::Rails::Engine
+
+    isolate_namespace DifficultCustomer
+
```
remove

rlogic commented on lib/difficult_customer/engine.rb:
```diff
@@ -0,0 +1,9 @@
+# .
+module DifficultCustomer
+  # A Rails Engine used make the gem running.
+  class Engine < ::Rails::Engine
```
once again because I don't remember: why do we need engine?

mpptx replied:
Afaik an engine is necessary to make the host app see the files in app/, lib/, etc.

rlogic commented on lib/difficult_customer/helper.rb:
```diff
@@ -0,0 +1 @@
+../app/helpers/difficult_customer_helper.rb
```
why helpers aren't defined here?

mpptx replied:
That allows us to load the gem into app by placing ```= difficult_customer```in the layout file.

rlogic commented on lib/generators/difficult_customer/install_generator.rb:
```diff
@@ -0,0 +1,14 @@
+      def copy_config
+        template 'difficult_customer.yml', 'config/difficult_customer.yml'
+      end
+
```
remove

rlogic commented on lib/generators/difficult_customer/install_generator.rb:
```diff
@@ -0,0 +1,14 @@
+  module Generators
+    # This generator copies example config into config/.
+    class InstallGenerator < Rails::Generators::Base
+
```
remove

rlogic reviewed (commented):
Few things to fix until approved

mpptx committed:
CR Fixes - Removing .ruby*, Gemfile amendments.

mpptx committed:
" -> ' in gemspec.

mpptx committed:
Removing Gemfile.lock.

mpptx committed:
Removal of compiled gem.

mpptx committed:
Including Gemfile.lock in .gitignore, passing parameters to partials.

mpptx committed:
CR Fixes - `random_dom_ids`, `random_string` and their improvements.

mpptx committed:
Adding more intelligent way to display notifications.

mpptx committed:
Refactoring:
- Adding Message, MessageDisplay, Defaults classes,
- Presenting messages with 'high' priority first,
- Presenting messages from the specific time ranges.
